### PR TITLE
fix issue with fitting 2D models with constraints and weights

### LINKED ISSUE #6403: Possible bug in fitting with multidimensional arrays
This is a bug report received via feedback@astropy.org - unfortunately the email of the submitter no longer works, but I'll put this here in case the bug is still present:

---

I have juste upgraded to astropy 1.3.2

in modeling.fitting, line 600:

```python
full_deriv = (np.ravel(weights) * np.array(model.fit_deriv(x, y, *model.parameters)).T).T
```

gives me an error, when using the LevMarLSQFitter(), with Gaussian2D models, and weights.
For instance, with XX, YY, im and W being all 32x32 numpy array:

```python
    fitter=fitting.LevMarLSQFitter()
    fit = fitter(myGaussianModel, XX, YY, im, weights=W)

    /home/samuel/Apps/anaconda2/lib/python2.7/site-packages/astropy/modeling/fitting.py in _wrap_deriv(params, model, weights, x, y, z)
        598                 full_deriv = np.ravel(weights) * np.array(model.fit_deriv(x, *model.parameters))
        599             else:
    --> 600                 full_deriv = (np.ravel(weights) * np.array(model.fit_deriv(x, y, *model.parameters)).T).T
        601                                

    ValueError: operands could not be broadcast together with shapes (1024,) (32,32,6)
```

In astropy 1.1.2, this line was different:

```python
	full_deriv = np.array(model.fit_deriv(x, y, *model.parameters))
```
and I didn't get any error.

### PR DESCRIPTION
**Pull request by nden**
Closes #6403

### CODE REVIEW

**Comment by nden** [comment]
This needs a review in order to be merged. Anyone?

**Comment by pllim** [comment]
Since @astrofrog reported the original issue, I assigned him as the reviewer.

**Comment by bsipocz** [comment]
@astrofrog - could you review this please, or remilestone to 2.0.4?

File: astropy/modeling/fitting.py [hunk 1]
```diff
@@ -729,11 +729,20 @@ def _wrap_deriv(params, model, weights, x, y, z=None):
...
 
         if any(model.fixed.values()) or any(model.tied.values()):
-
+            # update the parameters with the current values from the fitter
+            _fitter_to_model_params(model, params)
             if z is None:
-                full_deriv = np.ravel(weights) * np.array(model.fit_deriv(x, *model.parameters))
+                full = np.ravel(weights) * np.array(model.fit_deriv(x, *model.parameters))
+                if not model.col_fit_deriv:
+                    full_deriv = np.ravel(weights) * full.T
```
Comment (astrofrog) [inline astropy/modeling/fitting.py]: @nden - ``full_deriv`` will include ``np.ravel(weights)`` twice - is that normal? (that appears to be a change from before)

Comment (nden) [reply astropy/modeling/fitting.py]: @astrofrog Good catch! That would have applied weights twice. Thanks.

**Comment by nden** [comment]
This missed the `2.0.3` release. I hope it can make into `3.0`. As a start I'm offering a doughnut to anyone with write permissions who reviews it.

**Comment by pllim** [comment]
@nden, do you mind rebase to resolve conflict (and maybe squash while you are at it)? I trust your tests, so once conflict is gone, I'll approve it. (Yay, donut!)

**Comment by bsipocz** [comment]
Thanks @nden!

**Comment by nden** [comment]
:doughnut:

**Pull request merged**